src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerDriver.CompilationData.cs

```diff
@@ -64,31 +64,48 @@ public bool RemoveCachedSemanticModel(SyntaxTree tree)
 
             internal DeclarationAnalysisData GetOrComputeDeclarationAnalysisData(
                 SyntaxReference declaration,
-                Func<Func<DeclarationAnalysisData>, DeclarationAnalysisData> computeDeclarationAnalysisData,
+                Func<Func<DeclarationAnalysisDataBuilder>, DeclarationAnalysisDataBuilder> computeDeclarationAnalysisData,
                 bool cacheAnalysisData)
             {
-                if (!cacheAnalysisData)
+                DeclarationAnalysisDataBuilder dataBuilder = null;
+                try
                 {
-                    return computeDeclarationAnalysisData(_declarationAnalysisDataPool.Allocate);
-                }
+                    if (!cacheAnalysisData)
+                    {
+                        dataBuilder = computeDeclarationAnalysisData(_declarationAnalysisDataBuilderPool.Allocate);
+                        return DeclarationAnalysisData.CreateFrom(dataBuilder);
+                    }
 
-                DeclarationAnalysisData data;
-                lock (_declarationAnalysisDataMap)
-                {
-                    if (_declarationAnalysisDataMap.TryGetValue(declaration, out data))
+                    DeclarationAnalysisData data;
+                    lock (_declarationAnalysisDataMap)
                     {
-                        return data;
+                        if (_declarationAnalysisDataMap.TryGetValue(declaration, out data))
+                        {
+                            return data;
+                        }
                     }
-                }
 
-                data = computeDeclarationAnalysisData(_declarationAnalysisDataPool.Allocate);
+                    dataBuilder = computeDeclarationAnalysisData(_declarationAnalysisDataBuilderPool.Allocate);
```

Comment:
so, originally pool was there to reduce DeclarationAnalysisData type (and Lists in it). but now we always create those (including immutable arrays)

this now introduces new type builder which is pooled (so its builders as well) but all these are new types. so, pooling this doesnt include new allocations but also doesnt reduce any.

so, at the end, it is same as not having pool at all since we always create DeclarationAnalysisData (and its array).

if we are going this approach, I think we should just remove the pool. 

by the way, the pool is not static. and I think we know doesnt share same analyzer driver between compilation. so do we actually need this clear declaration analysis data anymore?

Reply:
I would rather make sure AnalyzerDriver state is not shared between compilation with analyzer in host and deprecate ClearCacheXXX methods.

Reply:
Sure, let me get rid of this pool as well and make it cleaner.